ocs_ci/utility/memory.py

```diff
@@ -0,0 +1,281 @@
+# -*- coding: utf8 -*-
+"""
+Module for memory related util functions.
+"""
+
+import os
+import logging
+import tempfile
+import numpy as np
+import pandas as pd
+from psutil import Process, ZombieProcess, NoSuchProcess
+from psutil._common import bytes2human
+from ocs_ci.ocs import constants
+from threading import Timer
+
+from ocs_ci.utility.utils import get_testrun_name
+
+current_factory = logging.getLogRecordFactory()
+log = logging.getLogger(__name__)
+
+
+class MemoryMonitor(Timer):
+    """
+    class to implement threading.Timer running func in a background
+    """
+
+    def run(self):
+        while not self.finished.wait(self.interval):
+            self.function(*self.args, **self.kwargs)
+
+
+consumed_ram_log = []
+_columns_df = ["pid", "name", "ts", "rss", "vms", "status"]
+_df = pd.DataFrame(columns=_columns_df)
+mon: MemoryMonitor
+_mem_csv: str
+
+
+def _get_memory_per_process():
+    """
+    Function to add memory rss and vms of current process and all subprocesses to dataframe obj (_df)
+    """
+    proc = Process(os.getpid())
+    _rec_memory(proc)
+    children = proc.children(recursive=True)
+    for child in children:
+        _rec_memory(child)
+    del proc
+
+
+def _rec_memory(proc: Process):
+    """
+    Helper func to update _df DataFrame object with proc stats, accordingly
+    to structure: "pid", "name", "ts", "rss", "vms", "status"
+    """
+    try:
+        global _df
+        _df = pd.concat(
+            [
+                _df,
+                pd.DataFrame(
+                    [
+                        [
+                            proc.pid,
+                            proc.name(),
+                            pd.Timestamp.now().strftime("%Y-%m-%d %X"),
+                            get_consumed_ram(proc),
+                            get_consumed_virt_mem(proc),
+                            proc.status(),
+                        ]
+                    ],
+                    columns=_columns_df,
+                ),
+            ]
+        )
+    except ZombieProcess:
+        log.exception("Cannot access ZombieProcess. Recording skipped")
+    except NoSuchProcess:
+        log.exception("Process is unavailable. Recording skipped")
+    except ValueError:
+        pass
```

Comment:
why are we passing ValueError? won't that lead to a blank field?

Reply:
it is ok to have 1 empty line because we're pulling data each 3 sec during long hours (though I did not receive it through dozens of runs). Most important to not fail the monitor during test run. Also we're processing empty lines with def catch_empty_mem_df